File: packages/python/readme_metrics/VerifyWebhook.py
import hmac
import json
import typing as t
from datetime import datetime, timedelta


class VerificationError(Exception):
    pass


class VerifyWebhook:
    def __init__(self, body: dict, signature: t.Optional[str], secret: str):
        if signature is None:
            raise VerificationError("Missing Signature")

        parsed_input = dict(
            (x.strip(), y.strip())
            for x, y in (element.split("=") for element in signature.split(","))
        )

        time = parsed_input["t"]

        if datetime.now() - datetime.fromtimestamp(
            int(parsed_input["t"]) / 1000
        ) > timedelta(minutes=30):
            raise VerificationError("Expired Signature")

        unsigned = time + "." + json.dumps(body, separators=(",", ":"))
        verify_signature = hmac.new(
            secret.encode("utf-8", errors="ignore"),
            unsigned.encode("utf8"),
            "sha256",
        ).hexdigest()
        readme_signature = parsed_input["v0"]
        if (
            hmac.compare_digest(
                verify_signature.encode(), readme_signature.encode("utf8")
            )
            is False
        ):
            raise VerificationError("Invalid Signature")

File: packages/node/src/lib/verify-webhook.ts
import crypto from 'crypto';

interface WebhookBody {
  email: string;
}

export default function verifyWebhook(body: WebhookBody, signature: string, secret: string): WebhookBody {
  if (!signature) throw new Error('Missing Signature');

  // Inspired by stripe-node
  // https://github.com/stripe/stripe-node/blob/4e82ccafda2017654ac264c070e7ebfa0e662fcd/lib/Webhooks.js#L240-L258
  const expectedScheme = 'v0';
  /* eslint-disable no-param-reassign */
  const { time, readmeSignature } = signature.split(',').reduce(
    (accum, item) => {
      const kv = item.split('=');

      if (kv[0] === 't') {
        accum.time = Number(kv[1]);
      }

      if (kv[0] === expectedScheme) {
        accum.readmeSignature = kv[1];
      }

      return accum;
    },
    {
      time: -1,
      readmeSignature: '',
    },
  );
  /* eslint-enable no-param-reassign */

  // Make sure timestamp is recent to prevent replay attacks
  const THIRTY_MIN = 30 * 60 * 1000;
  if (new Date().getTime() - new Date(+time).getTime() > THIRTY_MIN) {
    throw new Error('Expired Signature');
  }

  // Verify the signature is valid
  const unsigned = `${time}.${JSON.stringify(body)}`;
  const hmac = crypto.createHmac('sha256', secret);
  const verifySignature = hmac.update(unsigned).digest('hex');
  if (verifySignature !== readmeSignature) {
    throw new Error('Invalid Signature');
  }

  return body;
}
